File: backend/src/main/java/com/hgw/gestionale/repairdetails/controller/RepairDetailsController.java
package com.hgw.gestionale.repairdetails.controller;

import com.hgw.gestionale.repairdetails.dto.AddMessageRequest;
import com.hgw.gestionale.repairdetails.dto.CreateRepairDetailsRequest;
import com.hgw.gestionale.repairdetails.dto.RepairDetailsResponse;
import com.hgw.gestionale.repairdetails.dto.RepairMessageResponse;
import com.hgw.gestionale.repairdetails.dto.UpdateRepairDetailsRequest;
import com.hgw.gestionale.repairdetails.service.RepairDetailsService;
import com.hgw.gestionale.repairdetails.service.RepairMessageService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping(value = "/api/repairs/{repairId}/details", produces = MediaType.APPLICATION_JSON_VALUE)
@RequiredArgsConstructor
@Slf4j
public class RepairDetailsController {
    private final RepairDetailsService repairDetailsService;
    private final RepairMessageService repairMessageService;

    @GetMapping
    public RepairDetailsResponse getDetails(@PathVariable Long repairId) {
        log.info("RepairDetailsController.get repairId={}", repairId);
        return repairDetailsService.getByRepairId(repairId);
    }

    @PutMapping
    public RepairDetailsResponse updateDetails(
            @PathVariable Long repairId,
            @Valid @RequestBody UpdateRepairDetailsRequest request
    ) {
        log.info("RepairDetailsController.update repairId={}", repairId);
        return repairDetailsService.update(repairId, request);
    }

    @DeleteMapping
    public void deleteDetails(@PathVariable Long repairId) {
        log.info("RepairDetailsController.delete repairId={}", repairId);
        repairDetailsService.delete(repairId);
    }

    @PostMapping("/messages")
    @ResponseStatus(HttpStatus.CREATED)
    public RepairMessageResponse addMessage(
            @PathVariable Long repairId,
            @Valid @RequestBody AddMessageRequest request,
            Authentication authentication
    ) {
        log.info("RepairDetailsController.addMessage repairId={}", repairId);
        return repairMessageService.addMessage(repairId, request, authentication.getName());
    }

    @DeleteMapping("/messages/{messageId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteMessage(@PathVariable Long repairId, @PathVariable Long messageId) {
        log.info("RepairDetailsController.deleteMessage repairId={} messageId={}", repairId, messageId);
        repairMessageService.deleteMessage(messageId);
    }
}

File: backend-node/src/controller/repairDetailsController.ts
import {
    Body,
    Controller,
    Delete,
    Get,
    OperationId,
    Patch,
    Path,
    Post,
    Put,
    Request,
    Route,
    Security,
    Tags,
} from "tsoa";
import {
    RepairDetailsResponse,
    UpdateRepairDetailsRequest,
    UpdateDataConsegnaRequest,
    AddMessageRequest,
    UpdateMessageRequest,
    RepairMessageResponse,
} from "../dto/repairDetails.dto";
import { RepairDetailsService } from "../service/repairDetails.service";
import { AccessTokenPayload } from "../common/token";
import express from "express";

const repairDetailsService = new RepairDetailsService();

@Route("repairs/{repairId}/details")
@Tags("RepairDetails")
@Security("bearerAuth")
export class RepairDetailsController extends Controller {
    @Get("/")
    @OperationId("getRepairDetails")
    public async getRepairDetails(@Path() repairId: string): Promise<RepairDetailsResponse> {
        return repairDetailsService.get(repairId);
    }

    @Put("/")
    @OperationId("updateRepairDetails")
    public async updateRepairDetails(
        @Path() repairId: string,
        @Body() body: UpdateRepairDetailsRequest
    ): Promise<RepairDetailsResponse> {
        return repairDetailsService.update(repairId, body);
    }

    @Patch("/data-consegna")
    @OperationId("updateDataConsegna")
    public async updateDataConsegna(
        @Path() repairId: string,
        @Body() body: UpdateDataConsegnaRequest
    ): Promise<RepairDetailsResponse> {
        return repairDetailsService.updateDataConsegna(repairId, body);
    }

    @Delete("/")
    @OperationId("deleteRepairDetails")
    public async deleteRepairDetails(@Path() repairId: string): Promise<void> {
        return repairDetailsService.delete(repairId);
    }

    @Post("/messages")
    @OperationId("addRepairMessage")
    public async addRepairMessage(
        @Path() repairId: string,
        @Body() body: AddMessageRequest,
        @Request() request: express.Request
    ): Promise<RepairMessageResponse> {
        const autore = ((request as unknown as { user: AccessTokenPayload }).user).sub!;
        return repairDetailsService.addMessage(repairId, body, autore);
    }

    @Put("/messages/{messageId}")
    @OperationId("updateRepairMessage")
    public async updateRepairMessage(
        @Path() repairId: string,
        @Path() messageId: number,
        @Body() body: UpdateMessageRequest,
        @Request() request: express.Request
    ): Promise<RepairMessageResponse> {
        const autore = ((request as unknown as { user: AccessTokenPayload }).user).sub!;
        return repairDetailsService.updateMessage(repairId, messageId, body, autore);
    }

    @Delete("/messages/{messageId}")
    @OperationId("deleteRepairMessage")
    public async deleteRepairMessage(
        @Path() repairId: string,
        @Path() messageId: number,
        @Request() request: express.Request
    ): Promise<void> {
        const autore = ((request as unknown as { user: AccessTokenPayload }).user).sub!;
        return repairDetailsService.deleteMessage(repairId, messageId, autore);
    }
}
